central/sensor/service/pipeline/nodeinventory/two_pipelines_test.go

```diff
@@ -0,0 +1,393 @@
+		enricher                  nodeEnricher.NodeEnricher
+		operations                []func(t *testing.T, np pipeline.Fragment, ninvp pipeline.Fragment) error
+		setUpMocks                func(t *testing.T, m *usedMocks)
+		wantNodeExists            bool
+		wantKernelVersionNode     string
+		wantKernelVersionNodeScan string
+	}{
+		"lone node inventory should not find the node in DB": {
+			operations: []func(t *testing.T, np pipeline.Fragment, ninvp pipeline.Fragment) error{
+				// Node-scan (node-inventory) for node1 arrives over the node-inventory pipeline
+				func(t *testing.T, np pipeline.Fragment, ninvp pipeline.Fragment) error {
+					return ninvp.Run(context.Background(), clusterID, createNodeInventoryMsg(nodeID, "v2"), nil)
+				},
+			},
+			setUpMocks: func(t *testing.T, m *usedMocks) {
+				m.nodeDatastore.EXPECT().GetNode(gomock.Any(), gomock.Eq(nodeID)).MinTimes(1).Return(nil, false, nil)
+			},
+			wantNodeExists:        false,
+			wantKernelVersionNode: "",
+		},
+		"node inventory arriving after node should result in data from the node being overwritten": {
+			operations: []func(t *testing.T, np pipeline.Fragment, ninvp pipeline.Fragment) error{
+				// Old node-scan for node1 arrives over the node pipeline
+				func(t *testing.T, np pipeline.Fragment, ninvp pipeline.Fragment) error {
+					return np.Run(context.Background(), clusterID, createNodeMsg(nodeID, "v1"), nil)
+				},
+				// New node-scan (node-inventory) for node1 arrives over the node-inventory pipeline
+				func(t *testing.T, np pipeline.Fragment, ninvp pipeline.Fragment) error {
+					return ninvp.Run(context.Background(), clusterID, createNodeInventoryMsg(nodeID, "v2"), nil)
+				},
+			},
+			setUpMocks: func(t *testing.T, m *usedMocks) {
+				gomock.InOrder(
+					// node arrives
+					m.clusterStore.EXPECT().GetClusterName(gomock.Any(), gomock.Eq(clusterID)).Times(1).Return(clusterID, true, nil),
+					m.cveDatastore.EXPECT().EnrichNodeWithSuppressedCVEs(gomock.Any()),
+					m.riskStorage.EXPECT().UpsertRisk(gomock.Any(), gomock.Any()).AnyTimes().Return(nil),
+					m.nodeDatastore.EXPECT().UpsertNode(gomock.Any(), gomock.Any()).AnyTimes().Return(nil),
+
+					// node inventory arrives
+					m.nodeDatastore.EXPECT().GetNode(gomock.Any(), gomock.Eq(nodeID)).AnyTimes().Return(nodeWithScore, true, nil),
+					m.cveDatastore.EXPECT().EnrichNodeWithSuppressedCVEs(gomock.Any()).AnyTimes().Return(),
+					m.riskStorage.EXPECT().UpsertRisk(gomock.Any(), gomock.Any()).AnyTimes().Return(nil),
+					m.nodeDatastore.EXPECT().UpsertNode(gomock.Any(), nodeWithScanWithKernelV2).AnyTimes().Return(nil),
+					// check what got stored in the DB
+					m.nodeDatastore.EXPECT().GetNode(gomock.Any(), gomock.Eq(nodeID)).AnyTimes().Return(nodeWithScanWithKernelV2, true, nil),
+				)
+			},
+			wantNodeExists:            true,
+			wantKernelVersionNode:     "v1",
+			wantKernelVersionNodeScan: "v2",
+		},
+		"node inventory arriving first should result in data from it being lost": {
```

Comment:
I think, the behavior is correct but that part needs some extra explanation.

We have two places where we store kernel version:
- A field in the `storage.Node` object
- An entry into the array of packages known as `storage.NodeScan`

Now, after a V1 scan arrives, both these location will be filled for the first time and stored in the DB.
Next, if a V2 scan arrive, then we update only the entry in `storage.NodeScan`. The field in the Node object remains intact and still holds the old value.
In the test above, I assumed that the v1 node scan (from K8s) emits kernel version `v1`, whereas the v2 node scan (from Node Inventory) emits kernel version `v2`. This is what the assertion checks here - in the node, we should still have `v1`, while in the NodeScan it should be updated to `v2`.

Reply:
@Maddosaurus let me know if you would like to capture any of my answer as a comment in code. Otherwise, I would keep the explanation in Github and merge the code as it is.

Reply:
Ah, I see. Thanks for the explanation! Feel free to go ahead as is.